File: preferences_dialog.py
"""Preferences dialog for fterm settings."""

from PyQt5.QtCore import Qt
from PyQt5.QtGui import QFont
from PyQt5.QtWidgets import (
    QDialog, QVBoxLayout, QHBoxLayout, QTabWidget, QWidget,
    QLabel, QComboBox, QSpinBox, QCheckBox, QLineEdit,
    QPushButton, QGroupBox, QFormLayout, QFontComboBox,
    QDialogButtonBox,
)
from themes import get_theme_names


class PreferencesDialog(QDialog):
    """Settings dialog with tabbed interface."""

    def __init__(self, settings, parent=None):
        super().__init__(parent)
        self._settings = settings
        self.setWindowTitle("Preferences")
        self.setMinimumSize(450, 400)

        layout = QVBoxLayout(self)

        # Tab widget
        self._tabs = QTabWidget()
        layout.addWidget(self._tabs)

        self._setup_general_tab()
        self._setup_appearance_tab()
        self._setup_session_tab()

        # Buttons
        buttons = QDialogButtonBox(
            QDialogButtonBox.Ok | QDialogButtonBox.Cancel | QDialogButtonBox.Apply
        )
        buttons.accepted.connect(self._accept)
        buttons.rejected.connect(self.reject)
        buttons.button(QDialogButtonBox.Apply).clicked.connect(self._apply)
        layout.addWidget(buttons)

    def _setup_general_tab(self):
        tab = QWidget()
        layout = QVBoxLayout(tab)

        # Shell group
        shell_group = QGroupBox("Shell")
        shell_layout = QFormLayout(shell_group)

        self._shell_edit = QLineEdit()
        self._shell_edit.setPlaceholderText("Leave empty for default ($SHELL)")
        self._shell_edit.setText(self._settings.get("shell", ""))
        shell_layout.addRow("Shell command:", self._shell_edit)

        layout.addWidget(shell_group)

        # Behavior group
        behavior_group = QGroupBox("Behavior")
        behavior_layout = QVBoxLayout(behavior_group)

        self._scrollback_spin = QSpinBox()
        self._scrollback_spin.setRange(100, 100000)
        self._scrollback_spin.setSingleStep(1000)
        self._scrollback_spin.setValue(self._settings.get("scrollback_lines", 10000))

        scroll_layout = QHBoxLayout()
        scroll_layout.addWidget(QLabel("Scrollback lines:"))
        scroll_layout.addWidget(self._scrollback_spin)
        behavior_layout.addLayout(scroll_layout)

        self._confirm_close_check = QCheckBox("Confirm close when processes are running")
        self._confirm_close_check.setChecked(self._settings.get("confirm_close_running", True))
        behavior_layout.addWidget(self._confirm_close_check)

        layout.addWidget(behavior_group)
        layout.addStretch()

        self._tabs.addTab(tab, "General")

    def _setup_appearance_tab(self):
        tab = QWidget()
        layout = QVBoxLayout(tab)

        # Theme group
        theme_group = QGroupBox("Theme")
        theme_layout = QFormLayout(theme_group)

        self._theme_combo = QComboBox()
        self._theme_combo.addItems(get_theme_names())
        current_theme = self._settings.get("theme", "Dark")
        idx = self._theme_combo.findText(current_theme)
        if idx >= 0:
            self._theme_combo.setCurrentIndex(idx)
        theme_layout.addRow("Color theme:", self._theme_combo)

        layout.addWidget(theme_group)

        # Font group
        font_group = QGroupBox("Font")
        font_layout = QFormLayout(font_group)

        self._font_combo = QFontComboBox()
        self._font_combo.setCurrentFont(
            QFont(self._settings.get("font_family", "Monospace"))
        )
        font_layout.addRow("Font family:", self._font_combo)

        self._font_size_spin = QSpinBox()
        self._font_size_spin.setRange(6, 72)
        self._font_size_spin.setValue(self._settings.get("font_size", 11))
        font_layout.addRow("Font size:", self._font_size_spin)

        layout.addWidget(font_group)

        # Cursor group
        cursor_group = QGroupBox("Cursor")
        cursor_layout = QFormLayout(cursor_group)

        self._cursor_style_combo = QComboBox()
        self._cursor_style_combo.addItems(["block", "underline", "bar"])
        current_style = self._settings.get("cursor_style", "block")
        idx = self._cursor_style_combo.findText(current_style)
        if idx >= 0:
            self._cursor_style_combo.setCurrentIndex(idx)
        cursor_layout.addRow("Cursor style:", self._cursor_style_combo)

        self._cursor_blink_check = QCheckBox("Blink cursor")
        self._cursor_blink_check.setChecked(self._settings.get("cursor_blink", True))
        cursor_layout.addRow(self._cursor_blink_check)

        layout.addWidget(cursor_group)

        # Padding group
        padding_group = QGroupBox("Terminal")
        padding_layout = QFormLayout(padding_group)

        self._padding_spin = QSpinBox()
        self._padding_spin.setRange(0, 32)
        self._padding_spin.setSingleStep(2)
        self._padding_spin.setValue(self._settings.get("terminal_padding", 4))
        padding_layout.addRow("Padding (px):", self._padding_spin)

        layout.addWidget(padding_group)
        layout.addStretch()

        self._tabs.addTab(tab, "Appearance")

    def _setup_session_tab(self):
        tab = QWidget()
        layout = QVBoxLayout(tab)

        session_group = QGroupBox("Session")
        session_layout = QVBoxLayout(session_group)

        self._auto_save_check = QCheckBox("Auto-save session on exit")
        self._auto_save_check.setChecked(self._settings.get("auto_save_session", True))
        session_layout.addWidget(self._auto_save_check)

        self._restore_check = QCheckBox("Restore previous session on startup")
        self._restore_check.setChecked(self._settings.get("restore_session", True))
        session_layout.addWidget(self._restore_check)

        layout.addWidget(session_group)
        layout.addStretch()

        self._tabs.addTab(tab, "Session")

    def _apply(self):
        """Apply settings without closing."""
        self._settings.set("shell", self._shell_edit.text())
        self._settings.set("scrollback_lines", self._scrollback_spin.value())
        self._settings.set("confirm_close_running", self._confirm_close_check.isChecked())
        self._settings.set("theme", self._theme_combo.currentText())
        self._settings.set("font_family", self._font_combo.currentFont().family())
        self._settings.set("font_size", self._font_size_spin.value())
        self._settings.set("cursor_style", self._cursor_style_combo.currentText())
        self._settings.set("cursor_blink", self._cursor_blink_check.isChecked())
        self._settings.set("terminal_padding", self._padding_spin.value())
        self._settings.set("auto_save_session", self._auto_save_check.isChecked())
        self._settings.set("restore_session", self._restore_check.isChecked())

    def _accept(self):
        """Apply and close."""
        self._apply()
        self.accept()

File: src/preferencesdialog.cpp
#include "preferencesdialog.h"
#include "thememanager.h"

#include <QDialogButtonBox>
#include <QDir>
#include <QFileDialog>
#include <QFormLayout>
#include <QGroupBox>
#include <QHBoxLayout>
#include <QPushButton>
#include <QTabWidget>
#include <QVBoxLayout>

PreferencesDialog::PreferencesDialog(Settings *settings, QWidget *parent)
    : QDialog(parent)
    , m_settings(settings)
{
    setWindowTitle("Preferences");
    setMinimumSize(450, 400);

    auto *layout = new QVBoxLayout(this);

    auto *tabs = new QTabWidget;
    tabs->addTab(createGeneralTab(), "General");
    tabs->addTab(createAppearanceTab(), "Appearance");
    tabs->addTab(createSessionTab(), "Session");
    layout->addWidget(tabs);

    auto *buttons = new QDialogButtonBox(
        QDialogButtonBox::Ok | QDialogButtonBox::Cancel | QDialogButtonBox::Apply);
    connect(buttons, &QDialogButtonBox::accepted, this, [this]() { apply(); accept(); });
    connect(buttons, &QDialogButtonBox::rejected, this, &QDialog::reject);
    connect(buttons->button(QDialogButtonBox::Apply), &QPushButton::clicked,
            this, &PreferencesDialog::apply);
    layout->addWidget(buttons);
}

QWidget *PreferencesDialog::createGeneralTab()
{
    auto *tab = new QWidget;
    auto *layout = new QVBoxLayout(tab);

    // Shell
    auto *shellGroup = new QGroupBox("Shell");
    auto *shellLayout = new QFormLayout(shellGroup);
    m_shellEdit = new QLineEdit(m_settings->get("shell").toString());
    m_shellEdit->setPlaceholderText("Leave empty for default ($SHELL)");
    shellLayout->addRow("Shell:", m_shellEdit);
    layout->addWidget(shellGroup);

    // Behavior
    auto *behaviorGroup = new QGroupBox("Behavior");
    auto *behaviorLayout = new QFormLayout(behaviorGroup);

    m_scrollbackSpin = new QSpinBox;
    m_scrollbackSpin->setRange(100, 100000);
    m_scrollbackSpin->setSingleStep(1000);
    m_scrollbackSpin->setValue(m_settings->get("scrollback_lines", 10000).toInt());
    behaviorLayout->addRow("Scrollback lines:", m_scrollbackSpin);

    m_confirmCloseCheck = new QCheckBox("Confirm close when processes are running");
    m_confirmCloseCheck->setChecked(m_settings->get("confirm_close_running", true).toBool());
    behaviorLayout->addRow(m_confirmCloseCheck);

    layout->addWidget(behaviorGroup);
    layout->addStretch();
    return tab;
}

QWidget *PreferencesDialog::createAppearanceTab()
{
    auto *tab = new QWidget;
    auto *layout = new QVBoxLayout(tab);

    // Theme
    auto *themeGroup = new QGroupBox("Theme");
    auto *themeLayout = new QFormLayout(themeGroup);
    m_themeCombo = new QComboBox;
    m_themeCombo->addItems(ThemeManager::themeNames());
    m_themeCombo->setCurrentText(m_settings->get("theme", "Dark").toString());
    themeLayout->addRow("Theme:", m_themeCombo);
    layout->addWidget(themeGroup);

    // Font
    auto *fontGroup = new QGroupBox("Font");
    auto *fontLayout = new QFormLayout(fontGroup);
    m_fontCombo = new QFontComboBox;
    m_fontCombo->setCurrentFont(QFont(m_settings->get("font_family", "Monospace").toString()));
    fontLayout->addRow("Family:", m_fontCombo);

    m_fontSizeSpin = new QSpinBox;
    m_fontSizeSpin->setRange(6, 72);
    m_fontSizeSpin->setValue(m_settings->get("font_size", 11).toInt());
    fontLayout->addRow("Size:", m_fontSizeSpin);
    layout->addWidget(fontGroup);

    // Cursor
    auto *cursorGroup = new QGroupBox("Cursor");
    auto *cursorLayout = new QFormLayout(cursorGroup);
    m_cursorStyleCombo = new QComboBox;
    m_cursorStyleCombo->addItems({"block", "underline", "bar"});
    m_cursorStyleCombo->setCurrentText(m_settings->get("cursor_style", "block").toString());
    cursorLayout->addRow("Style:", m_cursorStyleCombo);

    m_cursorBlinkCheck = new QCheckBox("Blink cursor");
    m_cursorBlinkCheck->setChecked(m_settings->get("cursor_blink", true).toBool());
    cursorLayout->addRow(m_cursorBlinkCheck);
    layout->addWidget(cursorGroup);

    // Terminal
    auto *termGroup = new QGroupBox("Terminal");
    auto *termLayout = new QFormLayout(termGroup);
    m_paddingSpin = new QSpinBox;
    m_paddingSpin->setRange(0, 32);
    m_paddingSpin->setSingleStep(2);
    m_paddingSpin->setValue(m_settings->get("terminal_padding", 4).toInt());
    termLayout->addRow("Padding:", m_paddingSpin);
    layout->addWidget(termGroup);

    layout->addStretch();
    return tab;
}

QWidget *PreferencesDialog::createSessionTab()
{
    auto *tab = new QWidget;
    auto *layout = new QVBoxLayout(tab);

    // Session
    auto *sessionGroup = new QGroupBox("Session");
    auto *sessionLayout = new QFormLayout(sessionGroup);

    m_autoSaveCheck = new QCheckBox("Auto-save session on exit");
    m_autoSaveCheck->setChecked(m_settings->get("auto_save_session", true).toBool());
    sessionLayout->addRow(m_autoSaveCheck);

    m_restoreCheck = new QCheckBox("Restore previous session on startup");
    m_restoreCheck->setChecked(m_settings->get("restore_session", true).toBool());
    sessionLayout->addRow(m_restoreCheck);
    layout->addWidget(sessionGroup);

    // SSH
    auto *sshGroup = new QGroupBox("SSH");
    auto *sshLayout = new QFormLayout(sshGroup);

    auto *keyLayout = new QHBoxLayout;
    m_sshKeyEdit = new QLineEdit(m_settings->get("default_ssh_identity_file").toString());
    m_sshKeyEdit->setPlaceholderText("Default identity file for all SSH sessions");
    auto *browseBtn = new QPushButton("Browse...");
    keyLayout->addWidget(m_sshKeyEdit);
    keyLayout->addWidget(browseBtn);
    sshLayout->addRow("Default SSH Key:", keyLayout);
    connect(browseBtn, &QPushButton::clicked, this, &PreferencesDialog::browseSSHKey);
    layout->addWidget(sshGroup);

    layout->addStretch();
    return tab;
}

void PreferencesDialog::apply()
{
    m_settings->set("shell", m_shellEdit->text().trimmed());
    m_settings->set("scrollback_lines", m_scrollbackSpin->value());
    m_settings->set("confirm_close_running", m_confirmCloseCheck->isChecked());
    m_settings->set("theme", m_themeCombo->currentText());
    m_settings->set("font_family", m_fontCombo->currentFont().family());
    m_settings->set("font_size", m_fontSizeSpin->value());
    m_settings->set("cursor_style", m_cursorStyleCombo->currentText());
    m_settings->set("cursor_blink", m_cursorBlinkCheck->isChecked());
    m_settings->set("terminal_padding", m_paddingSpin->value());
    m_settings->set("auto_save_session", m_autoSaveCheck->isChecked());
    m_settings->set("restore_session", m_restoreCheck->isChecked());
    m_settings->set("default_ssh_identity_file", m_sshKeyEdit->text().trimmed());
}

void PreferencesDialog::browseSSHKey()
{
    QString path = QFileDialog::getOpenFileName(
        this, "Select SSH Key", QDir::homePath() + "/.ssh");
    if (!path.isEmpty())
        m_sshKeyEdit->setText(path);
}
